### PR TITLE
[WIP] Improved parsing of pytorch models using torch.FX

### PR DESCRIPTION
**Pull request by JanFSchulte**
Current parsing of pytorch models uses a loop of the `named_modules` of the model (https://github.com/fastmachinelearning/hls4ml/blob/main/hls4ml/converters/pytorch_to_hls.py#L163). This has several disadvantages: 

- Captures only layers that are defined as members of the model class
- Can't infer the correct order of models
- Ignores other operations that are part of the `forward()` method of the model

In this PR, we propose to fix this by first created a graph representation of the model's `forward()` function using the symbolic tracing functionality of [https://pytorch.org/docs/stable/fx.html](torch.FX). Each operation in the `forward()` is represented by a node in the graph. Nodes can be of these types:
image

For example, for this model

```
class MyModuleConvRelu(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(3,3,3)
        
    def forward(self, x):
        y1 = self.conv(x)
        y = torch.relu(y1)
        y = y + y1
        y = torch.relu(y)
        return y
```
the resulting graph representation is 

```
graph():
    %x : [#users=1] = placeholder[target=x]
    %conv : [#users=2] = call_module[target=conv](args = (%x,), kwargs = {})
    %relu : [#users=1] = call_function[target=torch.relu](args = (%conv,), kwargs = {})
    %add : [#users=1] = call_function[target=operator.add](args = (%relu, %conv), kwargs = {})
    %relu_1 : [#users=1] = call_function[target=torch.relu](args = (%add,), kwargs = {})
    return relu_1
```

As the nodes in the graph follow the order of operations of the `forward()` function, we can then simply loop over them and parse each node into one node in the hls4ml model representation. For the parsing of the individual layers, existing code is used where available without significant changes. Functionality for more types of layers is also added by this PR. 

The types of layers currently understood by the parser are

- Linear
- Softmax
- Relu
- LeakyReLU
- ThresholdedReLU
- ELU
- PReLU
- Sigmoid
- BatchNorm2d
- BatchNorm1d'
- Batch_norm
- MaxPool1d
- MaxPool2d
- AvgPool1d
- AvgPool2d
- Add
- Subtract
- Multiply
- Average
- Maximum
- Minimum
- Concatenate
- Dot
- Conv1d
- Conv2d
- View
- Dropout
- Flatten
- Sequential

This PR also fixes https://github.com/fastmachinelearning/hls4ml/issues/409 

Changes are mostly confined to the frontend, but small changes are made to the backend in 

hls4ml/backends/fpga/passes/codegen.py
hls4ml/backends/vivado/passes/conv_stream.py
hls4ml/backends/vivado/passes/repack_stream.py

to handle the different order of inputs in pytorch compared to keras.

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)

## Tests

The new parsing was tested using 5-6 different pytorch model examples from around the web. In addition, I verified that the two example models for pytroch included with hls4ml get parsed successfully. 
 
## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/fastmachinelearning/hls4ml/blob/main/CONTRIBUTING.md).
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have installed and run `pre-commit` on the files I edited or added.
- [ ] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by JanFSchulte** [comment]
The baseline version of this new pytorch parser has reached a stable state now and the code is ready for review. This implementation of baseline support will then be used to built functionality for graph models down the road.

**Comment by JanFSchulte** [comment]
Closed because git history of this branch is irrecoverable messed up.
